homeassistant/components/vallox/config_flow.py

```diff
@@ -0,0 +1,129 @@
+"""Config flow for the Vallox integration."""
+from __future__ import annotations
+
+from contextlib import suppress
+import ipaddress
+import logging
+import re
+from typing import Any
+
+from vallox_websocket_api import Vallox
+from vallox_websocket_api.exceptions import ValloxApiException
+import voluptuous as vol
+
+from homeassistant import config_entries
+from homeassistant.const import CONF_HOST, CONF_NAME
+from homeassistant.core import HomeAssistant
+from homeassistant.data_entry_flow import FlowResult
+from homeassistant.exceptions import HomeAssistantError
+
+from .const import DEFAULT_NAME, DOMAIN
+
+_LOGGER = logging.getLogger(__name__)
+
+STEP_USER_DATA_SCHEMA = vol.Schema(
+    {
+        vol.Required(CONF_HOST): str,
+    }
+)
+
+VALLOX_CONNECTION_EXCEPTIONS = (
+    OSError,
+    ValloxApiException,
+)
+
+
+def host_valid(host: str) -> bool:
+    """Return True if hostname or IP address is valid."""
+    with suppress(ValueError):
+        if ipaddress.ip_address(host).version in [4, 6]:
+            return True
+
+    disallowed = re.compile(r"[^a-zA-Z\d\-]")
+    return all(x and not disallowed.search(x) for x in host.split("."))
+
```

Comment:
Previously we only allowed an ip address
`vol.Required(CONF_HOST): vol.All(ipaddress.ip_address, cv.string),`

Ideally we keep it just to an ip so we don't have to rely on dns.  If the integration supports one of the built-in discovery methods (https://developers.home-assistant.io/docs/creating_integration_manifest#zeroconf) we can update the ip in the config entry from discovery in a future PR.

Reply:
Indeed 🤔 Keeping it that way would also simplify things quite a bit. I guess it means that the whole `host_valid` method can be scrapped?

Reply:
Hm, does `vol.Required(CONF_HOST): vol.All(ipaddress.ip_address, cv.string)` work for config flow data schema? 🤔 I tried looking for prior art, but couldn't find anything, and if I change it and try to add a Vallox device, I get this exception:

```
2021-12-27 19:45:18 ERROR (MainThread) [aiohttp.server] Error handling request
Traceback (most recent call last):
  File "/usr/local/lib/python3.9/site-packages/aiohttp/web_protocol.py", line 435, in _handle_request
    resp = await request_handler(request)
  File "/usr/local/lib/python3.9/site-packages/aiohttp/web_app.py", line 504, in _handle
    resp = await handler(request)
  File "/usr/local/lib/python3.9/site-packages/aiohttp/web_middlewares.py", line 117, in impl
    return await handler(request)
  File "/workspaces/core/homeassistant/components/http/security_filter.py", line 60, in security_filter_middleware
    return await handler(request)
  File "/workspaces/core/homeassistant/components/http/forwarded.py", line 98, in forwarded_middleware
    return await handler(request)
  File "/workspaces/core/homeassistant/components/http/request_context.py", line 28, in request_context_middleware
    return await handler(request)
  File "/workspaces/core/homeassistant/components/http/ban.py", line 78, in ban_middleware
    return await handler(request)
  File "/workspaces/core/homeassistant/components/http/auth.py", line 181, in auth_middleware
    return await handler(request)
  File "/workspaces/core/homeassistant/components/http/view.py", line 137, in handle
    result = await result
  File "/workspaces/core/homeassistant/components/config/config_entries.py", line 130, in post
    return await super().post(request)
  File "/workspaces/core/homeassistant/components/http/data_validator.py", line 62, in wrapper
    result = await method(view, request, *args, **kwargs)
  File "/workspaces/core/homeassistant/helpers/data_entry_flow.py", line 86, in post
    result = self._prepare_result_json(result)
  File "/workspaces/core/homeassistant/components/config/config_entries.py", line 134, in _prepare_result_json
    return _prepare_config_flow_result_json(result, super()._prepare_result_json)
  File "/workspaces/core/homeassistant/components/config/config_entries.py", line 104, in _prepare_config_flow_result_json
    return prepare_result_json(result)
  File "/workspaces/core/homeassistant/helpers/data_entry_flow.py", line 45, in _prepare_result_json
    data["data_schema"] = voluptuous_serialize.convert(
  File "/usr/local/lib/python3.9/site-packages/voluptuous_serialize/__init__.py", line 40, in convert
    pval = convert(value, custom_serializer=custom_serializer)
  File "/usr/local/lib/python3.9/site-packages/voluptuous_serialize/__init__.py", line 58, in convert
    val.update(convert(validator, custom_serializer=custom_serializer))
  File "/usr/local/lib/python3.9/site-packages/voluptuous_serialize/__init__.py", line 113, in convert
    if issubclass(schema, Enum):
TypeError: issubclass() arg 1 must be a class
```

Reply:
Kept the `str` type for now, but renamed `host_valid` to `ip_valid` and simplified it.

Reply:
The schema has to be serialized to json for the frontend so you can't use complex validators in config flows.